### PR TITLE
Export curve25519-dalek `u32_backend` feature

### PR DESCRIPTION
**Pull request by antoniokdacunha**
## Motivation
The motivation for this feature export is two fold:
    1. Using the curve25519-dalek `u32_backend` feature yields ~10x perf improvement for 32 bit architectures.
    2. The `u32_backend` and `u64_backend` features of curve25519-dalek are not additive but mutually exclusive. As a result, projects with dependencies on dalek crates (e.g this crypto crate) must expose these feature flags so crates which depend on them (e.g JellyfishMerkle) can in turn pick which backend to use. This is trivial for `u64_backend` which is enabled by default, but requires manual work (e.g this diff) for `u32_backend`.

### Have you read the [Contributing Guidelines on pull requests]

Yes

## Test Plan

Tests with all features pass:
cargo test
cargo test --no-default-features --features vanilla
cargo test --no-default-features --features vanilla-u32

### CODE REVIEW

File: crypto/crypto/src/ed25519.rs
Comment (huitseeker) [inline crypto/crypto/src/ed25519.rs:454]: Ha! We can actually remove this check, the [issue](https://github.com/dalek-cryptography/ed25519-dalek/issues/126) was fixed by the recent 1.0.0 release and [libra-update to it](https://github.com/libra/libra/pull/5779).

File: crypto/crypto/Cargo.toml
Comment (huitseeker) [inline crypto/crypto/Cargo.toml:61]: 1. if you have  `vanilla-(curve|ed|x)25519-dalek/u(64|32)_backend`, then `vanilla-(curve|ed|x)25519-dalek` should be redundant,
2. could we normalize the feature names to `vanilla_u64` and `vanilla_u32`?

**Review by huitseeker — CHANGES_REQUESTED** [review]
- Could you tell me more on the scope and intended usage of this? I'm asking because `fiat-rust ` generates [32-bits arithmetic](https://github.com/mit-plv/fiat-crypto/tree/master/fiat-rust/src) and we can definitely thread it through our [curve25519](https://github.com/novifinancial/curve25519-dalek/tree/fiat3), [ed25519](https://github.com/novifinancial/ed25519-dalek/tree/fiat4) and [x25519](https://github.com/novifinancial/x25519-dalek/tree/fiat3) forks.
- this needs to adjust the [CI run](https://github.com/libra/libra/blob/3972847fd3f81105183692360a8ea2a205e36b0a/.circleci/config.yml#L535-L547) so that it keeps testing the crypto crate's with every combination of non-default features.

File: crypto/crypto/src/ed25519.rs
Comment (antoniokdacunha) [reply crypto/crypto/src/ed25519.rs:454]: Is the check you want removed the entire thing or just the `not(any(feature = "vanilla", feature = "vanilla-u32"))` clause?

**Comment by antoniokdacunha** [comment]
> * Could you tell me more on the scope and intended usage of this? I'm asking because `fiat-rust ` generates [32-bits arithmetic](https://github.com/mit-plv/fiat-crypto/tree/master/fiat-rust/src) and we can definitely thread it through our [curve25519](https://github.com/novifinancial/curve25519-dalek/tree/fiat3), [ed25519](https://github.com/novifinancial/ed25519-dalek/tree/fiat4) and [x25519](https://github.com/novifinancial/x25519-dalek/tree/fiat3) forks.
> * this needs to adjust the [CI run](https://github.com/libra/libra/blob/3972847fd3f81105183692360a8ea2a205e36b0a/.circleci/config.yml#L535-L547) so that it keeps testing the crypto crate's with every combination of non-default features.

I've updated the description to include more details on motivation. 

While using `fiat-rust` may yield the performance gains we are looking for, the fundamental motivation for this change is that the curve25519-dalek `u32_backend` and `u64_backend` features are mutually exclusive, and so this crypto crate fails to build when we try to turn on u32_backend downstream.

File: .circleci/config.yml
Comment (huitseeker) [inline .circleci/config.yml:548]: I'm afraid you'll need `no-default-features` here as well, otherwise you'll hit the exclusivity check on fiat + vanilla-u32.

File: crypto/crypto/src/ed25519.rs
Comment (huitseeker) [inline crypto/crypto/src/ed25519.rs:451]: :+1:

**Review by huitseeker — COMMENTED** [review]
This lgtm, modulo a quick fix on CircleCI configuration.

File: .circleci/config.yml
Comment (antoniokdacunha) [reply .circleci/config.yml:548]: `no-default-features` is included on last line of both test clauses.

LMK if I am missing something obvious here.

Comment (huitseeker) [reply .circleci/config.yml:548]: :facepalm:

**Review by huitseeker — APPROVED** [review]

**Comment by davidiw** [comment]
Thanks for the contribution. You need to sign the CLA before this can land: https://github.com/libra/libra/pull/6343#issuecomment-702617273

**Comment by davidiw** [comment]
Thanks for the contribution. You need to sign the CLA before this can land: https://github.com/libra/libra/pull/6343#issuecomment-702617273

**Comment by antoniokdacunha** [comment]
Hey @François Garillot. I hope you had a nice weekend! I reached out to davidiw who recommended I sign the individual contirbutor CLA, which I have done. 

However I'm still unable to merge this pull request with this error `You’re not authorized to merge this pull request`. How should I proceed?

**Review by davidiw — APPROVED** [review]

**Comment by davidiw** [comment]
/land

layers and layers of approvals

**Pull request closed without merging**
